File: parsers/dart/test/conformance_test.dart
import 'dart:io';

import 'package:synx/synx.dart';
import 'package:test/test.dart';

/// Replay every `.synx` in the shared corpus through this parser.
/// Skips when the corpus directory is absent.
void main() {
  test('corpus parses without error', () {
    final dir = _findCorpus();
    if (dir == null) {
      // Skip silently when corpus isn't bundled.
      return;
    }
    final files = Directory(dir)
        .listSync()
        .whereType<File>()
        .where((f) => f.path.endsWith('.synx'));
    var parsed = 0;
    var failed = 0;
    for (final f in files) {
      final text = f.readAsStringSync();
      final r = parseFull(text);
      if (r.root is SynxObj) {
        parsed++;
      } else {
        failed++;
        print('corpus ${f.path} did not yield an object');
      }
    }
    print('[corpus] parsed $parsed files, $failed failed');
    expect(failed, equals(0));
  });
}

String? _findCorpus() {
  final candidates = [
    'tests/conformance/cases',
    '../tests/conformance/cases',
    '../../tests/conformance/cases',
    '../../../tests/conformance/cases',
  ];
  for (final c in candidates) {
    final d = Directory(c);
    if (d.existsSync()) return d.absolute.path;
  }
  return null;
}

File: parsers/go/conformance_test.go
package synx

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// TestConformanceCorpus replays every `.synx` file in the shared corpus
// (../../tests/conformance/cases) through this parser. Missing corpus → skip.
func TestConformanceCorpus(t *testing.T) {
	dir := findCorpus()
	if dir == "" {
		t.Skip("conformance corpus not present in this checkout")
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Skipf("cannot read corpus: %v", err)
	}
	parsed, failed := 0, 0
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".synx") {
			continue
		}
		text, err := os.ReadFile(filepath.Join(dir, e.Name()))
		if err != nil {
			continue
		}
		r := Parse(string(text))
		if _, ok := r.Root.(ObjectValue); ok {
			parsed++
		} else {
			failed++
			t.Logf("corpus %q did not yield an object", e.Name())
		}
	}
	t.Logf("[corpus] parsed %d files, %d failed", parsed, failed)
	if failed > 0 {
		t.Fatalf("%d corpus files failed", failed)
	}
}

func findCorpus() string {
	candidates := []string{
		"tests/conformance/cases",
		"../tests/conformance/cases",
		"../../tests/conformance/cases",
		"../../../tests/conformance/cases",
	}
	cwd, _ := os.Getwd()
	for _, c := range candidates {
		p := filepath.Join(cwd, c)
		if fi, err := os.Stat(p); err == nil && fi.IsDir() {
			return p
		}
	}
	return ""
}
